tests/channel_test.c

```diff
@@ -281,8 +280,10 @@ static int s_test_channel_refcount(struct aws_allocator *allocator, void *ctx) {
 
     /* Release hold 2/2. The handler and channel should be destroyed. */
     aws_channel_release_hold(channel);
+    ASSERT_SUCCESS(aws_mutex_lock(&destroy_mutex));
```

Comment:
Reduce size of the critical section by moving this lock down.

Reply:
I don't think this mutex/condition variable need to exist anymore. We can just loop until destroy_called is true.

Before our 1st pass of tsan fixes, this test was broken because it didn't acquire the lock before changing the bool and signaling the condvar.

We changed the bool to be atomic. We're  using the mutex when waiting for the value and **not** using wait_pred(). But we're **not** using the mutex when changing the value and signaling the condvar. That means, if we set the bool and signal **before** the main thread waits for it, the main thread will never be woken up (except for spurious wakeups).

So a fully proper solution is either:
1) use an atomic, main thread just checks in in a while loop. No mutex or condvar required.
2) OR use the lock AND the condition variable when setting the bool, use a predicate when waiting for the bool. It's not necessary for the bool to be atomic at this point.

First off, I think this test had been timing dependent and could theoretically fail
they weren't using predicates so the

Reply:
I see. Yes agreed on the pathological case. I'll fix with (1) as we can remove some code with that option.

Reply:
Actually, thinking about this further, it is ok to keep the bool atomic and the condition variable if the main thread does a check with `aws_condition_variable_wait_pred`. In this case, the thread executing  destroy atomically updates the bool and signals the condition variable (no lock necessary). The main thread locks the mutex and calls `aws_condition_variable_wait_pred` with a predicate that atomically loads from the bool. The spurious wakeup case is purely due to the use of `aws_condition_variable_wait`.